src/abe/dippe.c

```diff
@@ -0,0 +1,455 @@
+
+    // gt^(alpha(T)A) (k x 1)
+    cfe_mat_GT_init(&tmp_mat_gt, gt_A.cols, gt_A.rows);
+    cfe_mat_GT_transpose(&tmp_mat_gt, &gt_A);
+    cfe_mat_GT_mul_vec(&(pk->gt_alpha_A), &tmp_mat_gt, &(sk->alpha));
+
+    // cleanup
+    cfe_mat_free(&tmp_mat);
+    cfe_mat_GT_free(&gt_A);
+    cfe_mat_GT_free(&tmp_mat_gt);
+}
+
+cfe_error cfe_dippe_encrypt(cfe_dippe_cipher *cipher, cfe_dippe *dippe, cfe_dippe_pub_key *pks[], size_t pks_len, cfe_dippe_policy_vector *pv, FP12_BN254 *msg) {
+    // check whether there is the same number of keys as policy vector components
+    if (pks_len != pv->len) {
+        return CFE_ERR_INSUFFICIENT_KEYS;
+    }
+
+    // temp vars
+    BIG_256_56 tmp_big;
+    ECP_BN254 tmp_g1;
+    FP12_BN254 tmp_gt;
+
+    // s (k x 1)
+    cfe_vec s;
+    cfe_vec_init(&s, dippe->assump_size);
+    cfe_uniform_sample_vec(&s, dippe->p);
+
+    // C0 = g1^As (k+1 x 1)
+    cfe_mat_G1_mul_vec(&(cipher->C0), &(dippe->g1_A), &s);
+
+    // g1^(W_A_s) (k+1 x 1)
+    cfe_vec_G1 g1_W_A_s;
+    cfe_vec_G1_init(&g1_W_A_s, (dippe->assump_size + 1));
+
+    // g1^(x_UA_s) (k+1 x 1)
+    cfe_vec_G1 g1_x_UA_s;
+    cfe_vec_G1_init(&g1_x_UA_s, (dippe->assump_size + 1));
+
+    // Ci = g1^(xU(T)+W(T))As (pol x k+1)
+    for (size_t m = 0; m < pks_len; m++) {
+        // g1^(W(T)As)
+        cfe_mat_G1_mul_vec(&g1_W_A_s, &(pks[m]->g1_W_A), &s);
+
+        // g1^(W(T)As)
+        for (size_t i = 0; i < (dippe->assump_size + 1); i++) {
+            ECP_BN254_inf(&(g1_x_UA_s.vec[i]));
+            for (size_t k = 0; k < dippe->assump_size; k++) {
+                ECP_BN254_copy(&tmp_g1, &(dippe->g1_UA.mat[i].vec[k]));
+                BIG_256_56_from_mpz(tmp_big, s.vec[k]);
+                ECP_BN254_mul(&tmp_g1, tmp_big);
+                ECP_BN254_add(&(g1_x_UA_s.vec[i]), &tmp_g1);
+            }
+            BIG_256_56_from_mpz(tmp_big, pv->pol.vec[m]);
+            ECP_BN254_mul(&(g1_x_UA_s.vec[i]), tmp_big);
+        }
+
+        for (size_t i = 0; i < (dippe->assump_size + 1); i++) {
+            ECP_BN254_copy(&(cipher->Ci.mat[m].vec[i]), &(g1_x_UA_s.vec[i]));
+            ECP_BN254_add(&(cipher->Ci.mat[m].vec[i]), &(g1_W_A_s.vec[i]));
+        }
+    }
+
+    // C_prime = m * gt^alpha(T)As
+    FP12_BN254_one(&(cipher->C_prime));
+    for (size_t m = 0; m < pks_len; m++) {
+        for (size_t k = 0; k < dippe->assump_size; k++) {
+            BIG_256_56_from_mpz(tmp_big, s.vec[k]);
+            FP12_BN254_pow(&tmp_gt, &(pks[m]->gt_alpha_A.vec[k]), tmp_big);
+            FP12_BN254_mul(&(cipher->C_prime), &tmp_gt);
+        }
+    }
+    FP12_BN254_mul(&(cipher->C_prime), msg);
+
+    // cleanup
+    cfe_vec_free(&s);
+    cfe_vec_G1_free(&g1_W_A_s);
+    cfe_vec_G1_free(&g1_x_UA_s);
+
+    return CFE_ERR_NONE;
+}
+
+cfe_error cfe_dippe_keygen(cfe_dippe_user_sec_key *usk, cfe_dippe *dippe, size_t usk_id, cfe_dippe_pub_key *pks[], size_t pks_len, cfe_dippe_sec_key *sk, cfe_dippe_attribute_vector *av, char gid[]) {
```

Comment:
I finally understood that the way you implemented `keygen` and `decrypt` is to work with policies (vectors) obtained as `threshold` or `conjunction` instructions. Wouldn't it be nicer if it was implemented to work with general vectors, i.e. as an inner-product predicate scheme. In this sense the function `cfe_dippe_attribute_vector_init` would serve only as a change from a list of attributes to a vector (like the function `cfe_dippe_threshold_policy_vector_init` in essence already does) . If you could change it, we would be happy, otherwise we will change it latter.

Reply:
Agree, I've removed the `cfe_dippe_attribute_vector` struct and replaced it with a `cfe_vec`.